### ISSUE TITLE
Unable to send mail : django : The SECRET_KEY setting must not be empty

### ISSUE THREAD

**Issue by matclab** [galexrt/docker-zulip#77]
I'm unable to send mail.

I tried running the manage.py send_test_mail as advertised in settings.py: 
```
docker run -it quay.io/galexrt/zulip:1.5.1-2 app:managepy send_test_email user@example.com
```

But it fails with:
```
Running manage.py ...
Traceback (most recent call last):
  File "/home/zulip/deployments/current/scripts/lib/log-management-command", line 15, in <module>
    if not os.path.exists(settings.MANAGEMENT_LOG_PATH):
  File "/home/zulip/deployments/current/zulip-venv/lib/python2.7/site-packages/django/conf/__init__.py", line 53, in __getattr__
    self._setup(name)
  File "/home/zulip/deployments/current/zulip-venv/lib/python2.7/site-packages/django/conf/__init__.py", line 41, in _setup
    self._wrapped = Settings(settings_module)
  File "/home/zulip/deployments/current/zulip-venv/lib/python2.7/site-packages/django/conf/__init__.py", line 116, in __init__
    raise ImproperlyConfigured("The SECRET_KEY setting must not be empty.")
django.core.exceptions.ImproperlyConfigured: The SECRET_KEY setting must not be empty.
Traceback (most recent call last):
  File "/home/zulip/deployments/current/manage.py", line 22, in <module>
    " ".join(sys.argv)])
  File "/usr/lib/python2.7/subprocess.py", line 540, in check_call
    raise CalledProcessError(retcode, cmd)
```

The logs of docker-compose.up says:
```
zulip_1      | Setting Zulip secrets ...
zulip_1      | Secrets already generated.
zulip_1      | Secret found for "rabbitmq_password".
zulip_1      | Secret found for "email_password".
zulip_1      | Secret found for "secret_key".
```

**Comment by matclab** [comment]
It appears that in the docker, there is no `/etc/zulip/zulip-secrets.conf`

**Comment by matclab** [comment]
Restarting from the beginning:

- docker-compose rm
- rm /opt/docker/zulip  -rf
- docker-compose up

I got the following log about secrets:
```
zulip_1      | Setting Zulip secrets ...
zulip_1      | Generating Zulip secrets ...
zulip_1      | Generated Camo config file /etc/default/camo
zulip_1      | Generated /etc/zulip/zulip-secrets.conf with auto-generated secrets!
zulip_1      | Secrets generation succeeded.
zulip_1      | Secret found for "rabbitmq_password".
zulip_1      | Empty secret for key "email_password".
zulip_1      | Secret found for "secret_key".
zulip_1      | Zulip secrets configuration succeeded.
```

But I got the following error in the logs:
```
zulip_1      | Setting up the database, schema and user ...
database_1   | ERROR:  permission denied to create role
database_1   | STATEMENT:  CREATE USER zulip;
zulip_1      | ERROR:  permission denied to create role
zulip_1      | ALTER ROLE
database_1   | ERROR:  database "zulip" already exists
database_1   | STATEMENT:  CREATE DATABASE zulip OWNER=zulip;
zulip_1      | ERROR:  database "zulip" already exists
database_1   | ERROR:  schema "zulip" already exists
database_1   | STATEMENT:  CREATE SCHEMA zulip AUTHORIZATION zulip;
zulip_1      | ERROR:  schema "zulip" already exists
zulip_1      | Creating tsearch_extras extension ...
database_1   | FATAL:  password authentication failed for user "postgres"
database_1   | DETAIL:  Connection matched pg_hba.conf line 94: "host all all 0.0.0.0/0 md5"
zulip_1      | psql: FATAL:  password authentication failed for user "postgres"
zulip_1      | Database structure recreated.
…
zulip_1      | + sudo rabbitmqctl delete_user zulip
zulip_1      | sudo: rabbitmqctl: command not found
zulip_1      | + true
zulip_1      | + sudo rabbitmqctl delete_user zulip
zulip_1      | sudo: rabbitmqctl: command not found
zulip_1      | + true
zulip_1      | + sudo rabbitmqctl delete_user guest
zulip_1      | sudo: rabbitmqctl: command not found
zulip_1      | + true
…
rabbitmq_1   | =WARNING REPORT==== 17-Mar-2017::13:01:44 ===
rabbitmq_1   | closing AMQP connection <0.554.0> (172.19.0.3:46241 -> 172.19.0.6:5672):
rabbitmq_1   | connection_closed_abruptly
zulip_1      | 2017-03-17 13:01:51,172 INFO reaped unknown pid 324
zulip_1      | 2017-03-17 13:01:52,412 INFO exited: zulip-postsetup-create_user (exit status 0; expected)
```

**Comment by galexrt** [comment]
@matclab I'm working on fix for the errors you get.
- [x] Fix `rabbitmqctl` is missing
- [ ] Look into the missing secrets file (making send test mail (or even all mail stuff) fail)

**Comment by galexrt** [comment]
@matclab The command you ran for testing sending mails is wrong.
The command needs to `exec` and not `run`. `run` would create a new container. `exec` executes the command in the existing container.
So the command needs to be:
```
docker exec -it CONTAINER_NAME entrypoint.sh app:managepy send_test_email user@example.com
```
Where `CONTAINER_NAME` is the name of the docker-zulip container.

**Comment by matclab** [comment]
Thank you very much @galexrt.
And of course you're right for the `docker exec` command. Dumb of me.
I'll probably try on monday with the `dev` tag.

**Comment by galexrt** [comment]
@matclab A big thanks to you too, as most users don't provide so much information for debugging as you did :+1: :)
If you should experience any other issues, please let me know and also please tell me if the `dev` tag worked for you, so I can release the new image version right away.

**Comment by galexrt** [comment]
Thanks for the update!
Could you please share the contents of the `/opt/docker/zulip/zulip/settings/etc-zulip/settings.py` file (remove all your passwords) so I can take a look at this better.

- [ ] Replace the "old" database image with the fork [galexrt/postgres-zuilp-tsearchextras](https://github.com/galexrt/postgres-zulip-tsearchextras).
- [ ] Fix the rabbitmqctl not connecting to the right host
- [ ] Investigate why the `SETTING_EMAIL_USE_TLS/SSL` are ignored?

**Comment by matclab** [comment]
Yes. Here it is, with my docker-compose.yml: [docker-compose.yml and /opt/docker/zulip/zulip/settings/etc-zulip/settings.py](https://gist.github.com/matclab/f2439ec36fe7a2873cd6aeee02709bb9)

**Comment by galexrt** [comment]
@matclab I'm looking into it.

**Comment by galexrt** [comment]
I fixed the issue with the `SETTING_*` vars not being correctly set or the wrong type being detected.
I'm working on the other issues.

**Comment by galexrt** [comment]
@matclab Have you tried logging in as the `ZULIP_USER_EMAIL` user?
The `ZULIP_USER_EMAIL` gets created on start of the image and get's granted admin privileges.

**Comment by galexrt** [comment]
I now have fixed all issues except the text search.
I'm going to look into the text search not working in #78.

**Comment by matclab** [comment]
Thanks, I'll give a try and let you know.

I indeed used ZULIP_USER_EMAIL to log in.
It's a shame that DISABLE_HTTPS will be removed, as I run zulib behind a nginx reverse proxy. I remember it caused problem due to the 443 port being in used, but there may be a mean to circumvent this difficulty.

Thanks a lot for the work you did.

**Comment by galexrt** [comment]
@matclab As stated the `DISABLE_HTTPS` needs to be removed as with it enabled (`DISABLE_HTTPS: true`), you can't even login anymore.

**Comment by matclab** [comment]
I still obtain:
```
ValueError: EMAIL_USE_TLS/EMAIL_USE_SSL are mutually exclusive, so only set one of those settings to True.
```
when running `docker exec -it 0978cd5ae349 entrypoint.sh app:managepy send_test_email mcl@example.com`. None of those variables appear in `/opt/docker/zulip/zulip/settings/etc-zulip/settings.py`.

I see no admin tab in the UI even logged with the `ZULIP_USER_EMAIL` which is also `SETTING_ZULIP_ADMINISTRATOR`.


Error seen in the log:
```
database_1   | ERROR:  role "zulip" already exists
database_1   | STATEMENT:  CREATE USER zulip;
database_1   | ERROR:  database "zulip" already exists
database_1   | STATEMENT:  CREATE SCHEMA zulip AUTHORIZATION zulip;
database_1   | FATAL:  password authentication failed for user "postgres"
zulip_1      | Setting up the database, schema and user ...
zulip_1      | ERROR:  role "zulip" already exists
zulip_1      | ALTER ROLE
rabbitmq_1   | 
zulip_1      | ERROR:  database "zulip" already exists
zulip_1      | ERROR:  schema "zulip" already exists
zulip_1      | Deleting user "zulip" ...
zulip_1      | Error: unable to connect to node rabbit@0978cd5ae349: nodedown
zulip_1      | 
rabbitmq_1   | =INFO REPORT==== 21-Mar-2017::14:45:56 ===
zulip_1      | DIAGNOSTICS
zulip_1      | ===========
zulip_1      | 
zulip_1      | nodes in question: [rabbit@0978cd5ae349]
zulip_1      | 
zulip_1      | hosts, their running nodes and ports:
rabbitmq_1   | Setting user tags for user 'zulip' to [administrator]
zulip_1      | - 0978cd5ae349: [{rabbitmqctl108,33859}]
rabbitmq_1   | 
zulip_1      | 
rabbitmq_1   | =INFO REPORT==== 21-Mar-2017::14:45:56 ===
zulip_1      | current node details:
zulip_1      | - node name: rabbitmqctl108@0978cd5ae349
zulip_1      | - home dir: /var/lib/rabbitmq
zulip_1      | - cookie hash: 9iz8f1I7VAkchHn0QmsCMw==
zulip_1      | Creating user "zulip" ...
rabbitmq_1   | =INFO REPORT==== 21-Mar-2017::14:45:56 ===
zulip_1      | Error: unable to connect to node rabbit@0978cd5ae349: nodedown
zulip_1      | 
rabbitmq_1   | Server startup complete; 0 plugins started.
zulip_1      | DIAGNOSTICS
zulip_1      | ===========
rabbitmq_1   | 
zulip_1      | 
zulip_1      | nodes in question: [rabbit@0978cd5ae349]
rabbitmq_1   | =INFO REPORT==== 21-Mar-2017::14:46:48 ===
zulip_1      | 
zulip_1      | hosts, their running nodes and ports:
rabbitmq_1   | accepting AMQP connection <0.307.0> (172.19.0.2:34347 -> 172.19.0.6:5672)
zulip_1      | - 0978cd5ae349: [{rabbitmqctl238,52693}]
zulip_1      | 
rabbitmq_1   | 
zulip_1      | current node details:
zulip_1      | - node name: rabbitmqctl238@0978cd5ae349
zulip_1      | - home dir: /var/lib/rabbitmq
zulip_1      | - cookie hash: 9iz8f1I7VAkchHn0QmsCMw==

```
To be sure I use the `dev` branch at commit 42b7b1d23ac8d652d1ff909b3fd71e0cae04b463, and run a `docker-compose up` form there after tweaking the variables in the docker-compose file. 

Sorry for those bad news

**Comment by galexrt** [comment]
@matclab I can't reproduce your error. Please delete your `settings` folder from your data volume on the host and, pull and rerun `docker-compose up`.

**Comment by matclab** [comment]
I haven't remove DISABLE_HTTPS for now, hope it is not part of the problem.
 ```
$ docker-compose rm
$ cd .. && git clone clone  https://github.com/galexrt/docker-zulip.git
$ cd docker-zulip && git checkout dev
$ docker-compose up
$ rm /opt/docker/zulip/ -rf
$ docker ps
a301f4cad3bf        rabbitmq:3.5.5                                        "/docker-entrypoin..."   50 seconds ago      Up 47 seconds       5672/tcp                        dockerzulip_rabbitmq_1
6eeffd21d39f        quay.io/sameersbn/memcached:latest                    "/sbin/entrypoint...."   50 seconds ago      Up 47 seconds       11211/tcp, 11211/udp            dockerzulip_memcached_1
8534c7be32aa        quay.io/galexrt/zulip:dev                             "/sbin/entrypoint...."   50 seconds ago      Up 47 seconds       443/tcp, 0.0.0.0:6378->80/tcp   dockerzulip_zulip_1
fa3e54c3f901        quay.io/sameersbn/redis:latest                        "/sbin/entrypoint.sh"    50 seconds ago      Up 48 seconds       6379/tcp                        dockerzulip_redis_1
852afefa9f15        quay.io/galexrt/postgres-zulip-tsearchextras:latest   "docker-entrypoint..."   50 seconds ago      Up 47 seconds       5432/tcp                        dockerzulip_database_1
```
I got same errors.

Last commit given by `git log` is : 6d3f30d0c8a9944ce636cd9038cdf9a35e1da1f7

My docker-compose.yml file:
```
version: '2'
services:
  database:
    image: "quay.io/galexrt/postgres-zulip-tsearchextras:latest"
    environment:
      POSTGRES_DB: zulip
      POSTGRES_USER: zulip
      POSTGRES_PASSWORD: XXX
    volumes:
      - "/opt/docker/zulip/postgresql/data:/var/lib/postgresql:rw"
  memcached:
    image: "quay.io/sameersbn/memcached:latest"
    restart: always
  rabbitmq:
    image: "rabbitmq:3.5.5"
    hostname: zulip-rabbit
    restart: always
    environment:
        RABBITMQ_DEFAULT_USER: "zulip"
        RABBITMQ_DEFAULT_PASS: "XXX"
  redis:
    image: "quay.io/sameersbn/redis:latest"
    volumes:
      - "/opt/docker/zulip/redis:/var/lib/redis:rw"
  zulip:
    image: "quay.io/galexrt/zulip:dev"
    ports:
      - "6378:80"
      #- "443:443"
    environment:
      DB_HOST: "database"
      DB_USER: "zulip"
      DB_PASS: "bohJenguu6ae"
      SETTING_MEMCACHED_LOCATION: "memcached:11211"
      SETTING_RABBITMQ_HOST: "rabbitmq"
      SETTING_REDIS_HOST: "redis"
      SECRETS_email_password: ""
      SECRETS_rabbitmq_password: "XXX"
      SECRETS_secret_key: "XXX"
      SETTING_EXTERNAL_HOST: "chat.example.com"
      SETTING_ZULIP_ADMINISTRATOR: "mcl@example.com"
      SETTING_ADMIN_DOMAIN: "example.com"
      SETTING_NOREPLY_EMAIL_ADDRESS: "noreply@example.com"
      SETTING_DEFAULT_FROM_EMAIL: "Zulip <noreply@example.com>"
      SETTING_EMAIL_HOST: "localhost"
      SETTING_EMAIL_HOST_USER: ""
      SETTING_EMAIL_PORT: "587"
      SETTING_EMAIL_USE_SSL: "False"
      SETTING_EMAIL_USE_TLS: "True"
      SETTING_ALLOWED_HOSTS: "['chat.example.com']"
      ZULIP_AUTH_BACKENDS: "EmailAuthBackend"
      ZULIP_USER_CREATION_ENABLED: "True"
      ZULIP_USER_EMAIL: "mcl@example.com"
      ZULIP_USER_PASS: "zulip"
      ZULIP_USER_DOMAIN: "example.com"
      ZULIP_USER_FULLNAME: "Zulip Example User"
      DISABLE_HTTPS: "True"
    volumes:
      - "/opt/docker/zulip/zulip:/data:rw"
```

End of settings.py in /opt/zulip/*:
```
…
AUTHENTICATION_BACKENDS = ('zproject.backends.EmailAuthBackend',)
EXTERNAL_HOST = 'chat.example.com'
ALLOWED_HOSTS = ['chat.example.com']
NOREPLY_EMAIL_ADDRESS = 'noreply@example.com'
ZULIP_ADMINISTRATOR = 'mcl@example.com'
REDIS_HOST = 'redis'
RABBITMQ_HOST = 'rabbitmq'
DEFAULT_FROM_EMAIL = 'Zulip <noreply@example.com>'
MEMCACHED_LOCATION = 'memcached:11211'
ADMIN_DOMAIN = 'example.com'
```

And docker images hashes:
```
quay.io/galexrt/postgres-zulip-tsearchextras     latest              53e9a5fd7b4e        17 hours ago        273 MB
quay.io/galexrt/zulip                            dev                 d5c0c64adff9        4 days ago          959 MB
quay.io/sameersbn/redis                          latest              53f1377c5c57        7 weeks ago         202 MB
quay.io/sameersbn/memcached                      latest              695523c41d95        15 months ago       196 MB
rabbitmq                                         3.5.5               525b6bc90b2f        18 months ago       183 MB
```

**Comment by galexrt** [comment]
@matclab Can you please check the file `/home/zulip/deployments/current/zproject/settings.py` in the container?

**Comment by matclab** [comment]
Content seems right.
It ends with:
```
}
EMAIL_HOST = 'localhost'
EMAIL_USE_SSL = 'False'
EMAIL_USE_TLS = 'True'
EMAIL_PORT = '587'
```

Ah, but its python. There should be no quote around True/False.
Both were considered as being True.
Removing the quote inside the settings.py in the docker work.

Still to understand why they are generated with the quotes.

**Comment by galexrt** [comment]
@matclab Please post the full logs of the zulip container. There should be two lines like this:
```
zulip_1      | Setting key "EMAIL_USE_SSL", type "bool" in file "/home/zulip/deployments/current/zproject/settings.py".
zulip_1      | Setting key "EMAIL_USE_TLS", type "bool" in file "/home/zulip/deployments/current/zproject/settings.py".
```
That show as what type they got detected.

**Comment by matclab** [comment]
Indeed : 
```
zulip_1      | Setting key "EMAIL_USE_SSL", type "string" in file "/home/zulip/deployments/current/zproject/settings.py".
zulip_1      | Setting key "EMAIL_USE_TLS", type "string" in file "/home/zulip/deployments/current/zproject/settings.py".`

```

**Comment by galexrt** [comment]
@matclab Can you add the env var `DEBUG: "true"` to your `docker-compose.yml` and post the output then? (Please `docker-compose stop` and `docker-compose rm` the containers before running `docker-compose up` again).

**Comment by matclab** [comment]
Here it is : https://gist.github.com/matclab/2faa500b58a171069054caf2f6aa2afc

**Comment by galexrt** [comment]
Thanks for the gist!
I may have found a solution for your problem with the type detection (commit https://github.com/Galexrt/docker-zulip/commit/c1ddb53773d57a57bf815155d2401070e6543820).
Please repull the `dev` tag of the image and retest.

**Comment by matclab** [comment]
```
zulip_1      | + echo 'EMAIL_USE_SSL = False'
```
Seems good…

I have other problems (for the previous test I didn't do `docker-compose pull`, my bad). I'll keep you informed here.

**Comment by galexrt** [comment]
@matclab Did you try again with running `docker-compose pull` first?

**Comment by matclab** [comment]
@galexrt The quick try I did still show some errors about rabbitmq, but I didn't have time to really look at it. I'll probably manage to try again from start this week-end or on next monday.

**Comment by galexrt** [comment]
@matclab No problem. Please note you can ignore the `rabbitmqctl` errors.

**Comment by matclab** [comment]
@galexrt It does not look like there are critical errors in the log, however I can not test easily, because DISABLE_HTTPS does not work anymore, and zulip is only served over https in the docker. So I have to change my configuration (I had a nginx reverse proxy which was in charge of the SSL connexion toward the internet). Anyway I'll keep you informed if something does not work has expected.

**Comment by galexrt** [comment]
Thanks for the update!

@matclab As mentioned earlier I had to remove `DISABLE_HTTPS` as Zulip isn't working on http. Zulip only works over `https`.

**Comment by matclab** [comment]
Ok. For now I'm stucked with the fact that I shall configure nginx on a virtual domain to reverse proxy to the zulip docker `https`, but I must also respond with `https` to the `/.well-known` for letsencrypt certificate renewal… 

I'll post my (hopefully) success here for reference.

**Comment by galexrt** [comment]
@matclab You can try something like this (you would just need to add your reverse proxy stuff to it):
```
server {
    server_name zulip.example.com;
    listen 80;

    location /.well-known/acme-challenge {
        alias PATH_TO_ACME_CHALLENGES/.well-known/acme-challenge;
        location ~ /.well-known/acme-challenge/(.*) {
            default_type text/plain;
            try_files $uri =404;
        }
    }

    location / {
        return 301 https://zulip.example.com$request_uri;
    }
}

server {
    # Listen and server name
    server_name zulip.example.com;
    listen 443 ssl http2;
    listen [::]:443 ssl http2;

    # SSL stuff
    location / {
        # your proxy upstream stuff here
    }
}
```

**Comment by matclab** [comment]
Okay, that works. 
Thanks a lot !

- Full text search works.
- Mail sending works

However logging in with the user defined as  `ZULIP_USER_EMAIL` (which is also `SETTING_ZULIP_ADMINISTRATOR`) I still not see any administration item in the `settings` top-right menu.

**Comment by galexrt** [comment]
@matclab Can you try running the following command in the Zulip container:
```
/home/zulip/deployments/current/manage.py knight "YOUR_USER_EMAIL_TO_BE_ADMIN" -f
```
(Where `YOUR_USER_EMAIL_TO_BE_ADMIN` would be for example `example@example.com`)

**Comment by matclab** [comment]
Yep, with those commands, the administration tab is visible and operational. Thanks.

**Comment by galexrt** [comment]
@matclab Glad to help!
If there is anything else I can help you with. Please let me know!

**Comment by galexrt** [comment]
@matclab I will release the fix in the next few minuted so for new users this isn't required anymore.

**Comment by galexrt** [comment]
I just released tag `1.5.1-4` (branch `v1.5.1-4`).
The image build will take about 5-10 minutes. Please switch then to the `1.5.1-4` Docker image tag. Thank you!
